src/librustc_trans/trans/adt.rs

```diff
@@ -438,23 +469,33 @@ struct Case<'tcx> {
 /// This represents the (GEP) indices to follow to get to the discriminant field
 pub type DiscrField = Vec<usize>;
 
-fn find_discr_field_candidate<'tcx>(tcx: &ty::ctxt<'tcx>,
+fn find_discr_field_candidate<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
                                     ty: Ty<'tcx>,
-                                    mut path: DiscrField) -> Option<DiscrField> {
+                                    mut path: DiscrField) -> Option<(DiscrField, ValueRef)> {
+    let ref tcx = cx.tcx();
     match ty.sty {
-        // Fat &T/&mut T/Box<T> i.e. T is [T], str, or Trait
+        // Fat &T/&mut T/Box<T> i.e. T is [T], str, or Trait: null is forbidden
         ty::TyRef(_, ty::TypeAndMut { ty, .. }) | ty::TyBox(ty) if !type_is_sized(tcx, ty) => {
             path.push(FAT_PTR_ADDR);
-            Some(path)
+            Some((path, C_null(type_of::sizing_type_of(cx, ty))))
```

Comment:
@Yoric I may be misunderstanding the dialogue here, but @luqmana is not saying that we use a single byte as a discriminant. (I'm not exactly sure how you inferred that from his comments -- its not like we would dereference the pointer and look at the "byte" it points at to evaluate the discriminant here...)

All I think @luqmana  is saying is that if you want to build a null of the correct type, then you should use a null of type `*i8`, just because that happens to correspond to the types that trans is using for the two parts of a DST fat pointer.

I infer @luqmana found the type he is describing (as the shape of a DST fat pointer) by inspecting the source code of `sizing_type_of`.  If you don't want to build in that assumption, I imagine you could call `sizing_type_of` and then extract the appropriate type from the returned structure and pass that into `C_null`...